contracts/core/OriginCore.sol

```diff
@@ -0,0 +1,284 @@
+pragma solidity ^0.4.23;
+
+// Copyright 2018 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../OstInterface.sol";
+import "./OriginCoreConfig.sol";
+import "./OriginCoreInterface.sol";
+
+/**
+ * @title OriginCore is a proof of stake blockchain on Ethereum.
+ */
+contract OriginCore is OriginCoreInterface, OriginCoreConfig {
+
+    /* Events */
+
+    /** Emitted whenever a block is successfully reported. */
+    event BlockReported(
+        uint256 indexed height,
+        bytes32 indexed blockHash
+    );
+
+    /* Structs */
+
+    // TODO: We want to find a way to store the transaction root.
+    /** The header of an OSTblock. */
+    struct Header {
+        // TODO: is recomputation better than storing the header?
```

Comment:
if `.hashed` is only accessed in the `view`able function then it should definitely be calculated on the fly, as the gas is not deducted from the caller.  If later we should require it in a function where we do change state, then we should still default first to calculating it if it is not known from another source - main argument that the only expected hash in the header is the parent hash.

we will introduce the parenthash when we write the commit logic